File: slicing-ios/SlicingGame/Components/BasicViews/ImageView.swift
//
//  ImageView.swift
//  Basic view: an image view
//

import UIKit
import UniLayout
import JsonInflator

enum ImageStretchType: String {
    
    case none = "none"
    case fill = "fill"
    case aspectFit = "aspectFit"
    case aspectCrop = "aspectCrop"
    
    func toContentMode() -> UIView.ContentMode {
        switch self {
        case .fill:
            return .scaleToFill
        case .aspectFit:
            return .scaleAspectFit
        case .aspectCrop:
            return .scaleAspectFill
        case .none:
            return .center
        }
    }

}

class ImageView: UniImageView {
    
    // --
    // MARK: Viewlet integration
    // --

    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return ImageView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let imageView = object as? ImageView {
                // Apply image
                let stretchType = ImageStretchType(rawValue: convUtil.asString(value: attributes["stretchType"]) ?? "") ?? .none
                imageView.internalImageView.contentMode = stretchType.toContentMode()
                imageView.source = ImageSource.fromValue(value: attributes["source"])
                
                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: imageView, attributes: attributes)
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == ImageView.self
        }
        
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    private func setup() {
        clipsToBounds = true
        internalImageView.contentMode = .center
    }
    
    
    // --
    // MARK: Configurable values
    // --
    
    var source: ImageSource? {
        didSet {
            if let source = source {
                source.getImage(completion: { [weak self] image in
                    if self?.source === source {
                        self?.image = image
                    }
                })
            } else {
                image = nil
            }
        }
    }

}

File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/components/basicviews/ImageView.kt
package com.crescentflare.slicinggame.components.basicviews

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.components.utility.ImageSource
import com.crescentflare.slicinggame.components.utility.ViewletUtil
import com.crescentflare.unilayout.views.UniImageView


/**
 * Basic view: an image view
 */
open class ImageView : UniImageView {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return ImageView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is ImageView) {
                    // Apply image
                    val stretchType = StretchType.fromString(mapUtil.optionalString(attributes, "stretchType", ""))
                    obj.scaleType = stretchType.toScaleType()
                    obj.source = ImageSource.fromValue(attributes["source"])

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == ImageView::class
            }
        }
    }


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        // No implementation
    }


    // --
    // Configurable values
    // --

    var source: ImageSource? = null
        set(source) {
            field = source
            if (source != null) {
                source.getDrawable(context) {
                    if (field === source) {
                        setImageDrawable(it)
                    }
                }
            } else {
                setImageDrawable(null)
            }
        }


    // --
    // Scale type enum
    // --

    enum class StretchType(val value: String) {

        None("none"),
        Fill("fill"),
        AspectFit("aspectFit"),
        AspectCrop("aspectCrop");

        fun toScaleType(): ScaleType {
            return when(this) {
                Fill -> ScaleType.FIT_XY
                AspectFit -> ScaleType.FIT_CENTER
                AspectCrop -> ScaleType.CENTER_CROP
                else -> ScaleType.CENTER
            }
        }

        companion object {

            fun fromString(string: String?): StretchType {
                for (enum in values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return None
            }

        }

    }

}
